File: backend-haskell/src/BackendHaskell/Grpc/ProtoMapper.hs
-- | Domain.Task <-> Task.V1.Task(生成されたprotobufモジュール)の相互変換
module BackendHaskell.Grpc.ProtoMapper
  ( toProto
  , fromCreateRequest
  , fromUpdateRequest
  ) where

import Control.Lens ((&), (.~), (^.))
import Data.ProtoLens (defMessage)
import Data.ProtoLens.Labels ()
import qualified Data.Text as T
import Data.Time (Day, LocalTime (..), UTCTime (..), timeOfDayToTime)
import Data.Time.Clock.POSIX (utcTimeToPOSIXSeconds)
import Data.Time.Format (defaultTimeLocale, formatTime)

import BackendHaskell.Domain.Errors (TaskError (..))
import qualified BackendHaskell.Domain.Models as Dom
import BackendHaskell.Domain.Models (TaskInput (..))
import BackendHaskell.Domain.Validation (parseFinishedOn)
import qualified Proto.Google.Protobuf.Timestamp as PT
import Proto.Task.V1.Task
import Proto.Task.V1.Task_Fields

toProto :: Dom.Task -> Proto.Task.V1.Task.Task
toProto t =
  base
    & #id .~ fromIntegral (Dom.taskId t)
    & #name .~ Dom.taskName t
    & #status .~ Dom.taskStatusWire t
    & #finishedOn .~ dayToIso8601 (Dom.taskFinishedOn t)
    & #labels .~ map labelToProto (Dom.taskLabels t)
    & #createdAt .~ toTimestamp (Dom.taskCreatedAt t)
    & #updatedAt .~ toTimestamp (Dom.taskUpdatedAt t)
    & maybeSetDescription (Dom.taskDescription t)
  where
    base = defMessage :: Proto.Task.V1.Task.Task
    maybeSetDescription Nothing task = task
    maybeSetDescription (Just d) task = task & #description .~ d

labelToProto :: Dom.Label -> Proto.Task.V1.Task.Label
labelToProto l = defMessage & #id .~ fromIntegral (Dom.labelId l) & #name .~ Dom.labelName l

fromCreateRequest :: CreateTaskRequest -> Either TaskError TaskInput
fromCreateRequest req = do
  finishedOn <- parseFinishedOn (req ^. #finishedOn)
  Right
    TaskInput
      { tiName = req ^. #name
      , tiDescription = if req ^. #maybe'description == Nothing then Nothing else Just (req ^. #description)
      , tiStatusRaw = req ^. #status
      , tiFinishedOn = finishedOn
      , tiLabelIds = map fromIntegral (req ^. #labelIds)
      }

fromUpdateRequest :: UpdateTaskRequest -> Either TaskError TaskInput
fromUpdateRequest req = do
  finishedOn <- parseFinishedOn (req ^. #finishedOn)
  Right
    TaskInput
      { tiName = req ^. #name
      , tiDescription = if req ^. #maybe'description == Nothing then Nothing else Just (req ^. #description)
      , tiStatusRaw = req ^. #status
      , tiFinishedOn = finishedOn
      , tiLabelIds = map fromIntegral (req ^. #labelIds)
      }

dayToIso8601 :: Day -> T.Text
dayToIso8601 d = T.pack (formatTime defaultTimeLocale "%Y-%m-%d" d)

-- 【実機検証で確認する必要がある点、backend-python(FromDatetime、ナイーブなdatetimeをUTCとして
-- 扱う)/backend-java(JDBCのタイムゾーン変換バグ)/backend-elixir(NaiveDateTime->DateTime.from_naive!)
-- と同じ観点】LocalTimeはUTCの壁時計値そのものであることを前提に、明示的にUTC epoch秒へ
-- 変換してからgoogle.protobuf.Timestampへ渡す(タイムゾーン変換は一切発生しない)
toTimestamp :: LocalTime -> PT.Timestamp
toTimestamp lt = defMessage & #seconds .~ epochSeconds & #nanos .~ 0
  where
    -- LocalTime(タイムゾーン情報を持たない、UTCの壁時計値そのもの)を、実際のタイムゾーン
    -- 変換を一切行わずUTCTimeへそのまま組み直す(Day + TimeOfDayの構成要素をそのまま
    -- UTCTimeのDay + DiffTimeへ詰め替えるだけ)
    asUtc = UTCTime (localDay lt) (timeOfDayToTime (localTimeOfDay lt))
    epochSeconds = floor (utcTimeToPOSIXSeconds asUtc)

File: backend-elixir/lib/backend_elixir/grpc/proto_mapper.ex
defmodule BackendElixir.Grpc.ProtoMapper do
  @moduledoc "Domain.Task <-> Task.V1.Task(protobuf生成モジュール)の相互変換"

  alias BackendElixir.Domain.TaskInput
  alias BackendElixir.Domain.Validation

  def to_proto(task) do
    base = %Task.V1.Task{
      id: task.id,
      name: task.name,
      status: task.status_wire,
      finished_on: Date.to_iso8601(task.finished_on),
      labels: Enum.map(task.labels, fn label -> %Task.V1.Label{id: label.id, name: label.name} end),
      created_at: to_timestamp(task.created_at),
      updated_at: to_timestamp(task.updated_at)
    }

    if task.description do
      %{base | description: task.description}
    else
      base
    end
  end

  def from_create_request(%Task.V1.CreateTaskRequest{} = req) do
    with {:ok, finished_on} <- Validation.parse_finished_on(req.finished_on) do
      {:ok,
       %TaskInput{
         name: req.name,
         description: req.description,
         status_raw: req.status,
         finished_on: finished_on,
         label_ids: req.label_ids
       }}
    end
  end

  def from_update_request(%Task.V1.UpdateTaskRequest{} = req) do
    with {:ok, finished_on} <- Validation.parse_finished_on(req.finished_on) do
      {:ok,
       %TaskInput{
         name: req.name,
         description: req.description,
         status_raw: req.status,
         finished_on: finished_on,
         label_ids: req.label_ids
       }}
    end
  end

  # 【実機検証で確認する必要がある点、backend-python(FromDatetime、ナイーブなdatetimeをUTCとして
  # 扱う)/backend-java(JDBCのタイムゾーン変換バグ)と同じ観点】ここではNaiveDateTimeが
  # UTCの壁時計値そのものであることを前提に、明示的にUTCとしてDateTimeへ変換してから
  # Google.Protobuf.Timestampへ渡す(タイムゾーン変換は一切発生しない)
  defp to_timestamp(%NaiveDateTime{} = naive) do
    dt = DateTime.from_naive!(naive, "Etc/UTC")
    seconds = DateTime.to_unix(dt, :second)
    %Google.Protobuf.Timestamp{seconds: seconds, nanos: 0}
  end
end
